crates/worker/src/partition_processor_manager.rs

```diff
@@ -607,9 +615,34 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
             .await?
             .into_arc();
 
+        let mut tasks = JoinSet::new();
         for control_processor in control_processors.commands {
-            self.on_control_processor(control_processor, &partition_table)
+            let maybe_starting = self
+                .on_control_processor(control_processor, &partition_table)
                 .await?;
+
+            if let MaybeStarting::Starting(starting) = maybe_starting {
+                let tc = self.task_center.clone();
+                tasks.spawn(
+                    async move {
+                        tc.run_in_scope(
+                            "start-partition-processor",
+                            Some(control_processor.partition_id),
+                            starting,
+                        )
+                        .await
+                    }
+                    .instrument(debug_span!("start_partition_processor")),
+                );
+            }
+        }
+
+        while let Some(started) = tasks.join_next().await {
+            let state = started.unwrap()?;
+            self.invokers_status_reader
+                .push(state.key_range.clone(), state.status_reader.clone());
+            self.running_partition_processors
+                .insert(state.partition_id, state);
```

Comment:
Parallelizing the work speeds things up but it does not solve the underlying problem that we are blocking the `PartitionProcessorManager` from working on other messages. When I set the number of partitions to 128, then it takes roughly 9s for the `on_control_processors` call to complete on my machine. I suspect that there is some lock contention on the Tokio runtime when creating new runtimes + slow locks on Mac. On Linux the problem might be less severe because of faster futexes (maybe).

![localhost_16686_trace_f29b8571ef846f5254b6974f1a875652](https://github.com/user-attachments/assets/3fb44ad4-6a8e-4982-981b-bd37fca4b03f)

What I was initially wondering is whether we can restructure the code such that we don't have to await the completion of spawning the partition processors in `on_control_processors` but rather react to the completion in the top-most event loop or when operating on the partition processor again? I assume that this would boil down to a small state machine representing the state of the partition processor (spawning, running (leader or follower)).

Reply:
Maybe you could measure the time it takes on your Linux machine when starting a few more partition processors.

Reply:
If we could incorporate waiting for the updated metadata version in the not-blocking the event loop part, then this would be kick ass :-)

Reply:
Thank you @tillrohrmann  for clearing up my misunderstanding. My impression was that the problem is only during startup because of a slow start of the partition processors. I will work again on it